flashlight/app/imgclass/examples/ImageNetTransformer.cpp

```diff
@@ -0,0 +1,325 @@
+  // Setup distributed training
+  ////////////////////////
+  af::info();
+  if (FLAGS_distributed_enable) {
+    fl::ext::initDistributed(
+        FLAGS_distributed_world_rank,
+        FLAGS_distributed_world_size,
+        FLAGS_distributed_max_devices_per_node,
+        FLAGS_distributed_rndv_filepath);
+  }
+  const int worldRank = fl::getWorldRank();
+  const int worldSize = fl::getWorldSize();
+  const bool isMaster = (worldRank == 0);
+
+  af::setDevice(worldRank);
+  af::setSeed(worldSize);
+
+  auto reducer =
+      std::make_shared<fl::CoalescingReducer>(1.0 / worldSize, true, true);
+
+  //////////////////////////
+  //  Create datasets
+  /////////////////////////
+  FL_LOG_MASTER(INFO) << "Creating dataset";
+  // These are the mean and std for each channel of Imagenet
+  const std::vector<float> mean = {0.485, 0.456, 0.406};
+  const std::vector<float> std = {0.229, 0.224, 0.225};
+  const int randomResizeMax = 480;
+  const int randomResizeMin = 256;
+  const int randomCropSize = 224;
+  const float horizontalFlipProb = 0.5f;
+  // TransformDataset will apply each transform in a vector to the respective
+  // af::array. Thus, we need to `compose` all of the transforms so are each
+  // applied only to the image
+  ImageTransform trainTransforms = compose(
+      {// randomly resize shortest side of image between 256 to 480 for
+       // scale invariance
+       fl::ext::image::randomResizeTransform(randomResizeMin, randomResizeMax),
+       fl::ext::image::randomCropTransform(randomCropSize, randomCropSize),
+       fl::ext::image::normalizeImage(mean, std),
+       // Randomly flip image with probability of 0.5
+       fl::ext::image::randomHorizontalFlipTransform(horizontalFlipProb)});
+  ImageTransform valTransforms =
+      compose({// Resize shortest side to 256, then take a center crop
+               fl::ext::image::resizeTransform(randomResizeMin),
+               fl::ext::image::centerCropTransform(randomCropSize),
+               fl::ext::image::normalizeImage(mean, std)});
+
+  const int64_t batchSizePerGpu = FLAGS_data_batch_size;
+  const int64_t prefetchThreads = 10;
+  const int64_t prefetchSize = FLAGS_data_batch_size;
+  auto labelMap = getImagenetLabels(labelPath);
+  auto trainDataset = fl::ext::image::DistributedDataset(
+      imagenetDataset(trainList, labelMap, {trainTransforms}),
+      worldRank,
+      worldSize,
+      batchSizePerGpu,
+      prefetchThreads,
+      prefetchSize);
+  FL_LOG_MASTER(INFO) << "[trainDataset size] " << trainDataset.size();
+
+  auto valDataset = fl::ext::image::DistributedDataset(
+      imagenetDataset(valList, labelMap, {valTransforms}),
+      worldRank,
+      worldSize,
+      batchSizePerGpu,
+      prefetchThreads,
+      prefetchSize);
+  FL_LOG_MASTER(INFO) << "[valDataset size] " << valDataset.size();
+
+  //////////////////////////
+  //  Load model and optimizer
+  /////////////////////////
+  auto model = std::make_shared<fl::ext::image::ViT>(
```

Comment:
what about make patch size as a parameter? + addd linear layer if emb size is different from the the patches?